File: dotnet/src/MableBank.Infrastructure/ConsoleReporter.cs
using MableBank.Application;
using MableBank.Domain;

namespace MableBank.Infrastructure;

public static class ConsoleReporter
{
    public static string FormatReport(ProcessingReport report)
    {
        ArgumentNullException.ThrowIfNull(report);
        var lines = new List<string>
        {
            report.DryRun
                ? "Mable Bank Transfer Processing (dry run)"
                : "Mable Bank Transfer Processing",
            string.Empty,
            $"Transfers processed: {report.ProcessedCount}",
            $"Successful transfers: {report.SucceededCount}",
            $"Failed transfers: {report.FailedCount}",
        };

        if (report.SkippedCount > 0)
        {
            lines.Add($"Skipped transfers (already processed): {report.SkippedCount}");
        }

        if (report.DryRun)
        {
            lines.Add("No balances were mutated during dry run.");
        }

        lines.Add(string.Empty);
        lines.Add("Final balances:");

        foreach (var snapshot in report.FinalBalances)
        {
            lines.Add($"{snapshot.AccountNumber}: {MoneyFormatter.Format(snapshot.Balance)}");
        }

        var failures = report.FailedResults.OfType<TransferFailureResult>().ToList();
        if (failures.Count > 0)
        {
            lines.Add(string.Empty);
            lines.Add("Failed transfers:");
            for (var index = 0; index < failures.Count; index++)
            {
                var failure = failures[index];
                var instruction = failure.Instruction;
                lines.Add(
                    $"{index + 1}. From {instruction.FromAccountNumber} to {instruction.ToAccountNumber} " +
                    $"for {MoneyFormatter.Format(instruction.Amount)} - {failure.Message}");
            }
        }

        return string.Join(Environment.NewLine, lines);
    }
}

File: nodejs-typescript/src/infrastructure/ConsoleReporter.ts
import type { ProcessingReport } from '../application/ProcessingReport.js';
import { isTransferFailure } from '../domain/TransferResult.js';
import { formatMoney } from './MoneyFormatter.js';

export function formatReport(report: ProcessingReport): string {
  const lines: string[] = [
    report.dryRunMode
      ? 'Mable Bank Transfer Processing (dry run)'
      : 'Mable Bank Transfer Processing',
    '',
    `Transfers processed: ${String(report.processedCount)}`,
    `Successful transfers: ${String(report.succeededCount)}`,
    `Failed transfers: ${String(report.failedCount)}`,
  ];

  if (report.skippedCount > 0) {
    lines.push(`Skipped transfers (already processed): ${String(report.skippedCount)}`);
  }

  if (report.dryRunMode) {
    lines.push('No balances were mutated during dry run.');
  }

  lines.push('', 'Final balances:');
  for (const snapshot of report.finalBalances) {
    lines.push(`${snapshot.accountNumber.toString()}: ${formatMoney(snapshot.balance)}`);
  }

  const failures = report.failedResults.filter(isTransferFailure);
  if (failures.length > 0) {
    lines.push('', 'Failed transfers:');
    failures.forEach((result, index) => {
      const instruction = result.instruction;
      lines.push(
        `${String(index + 1)}. From ${instruction.fromAccountNumber.toString()} to ${instruction.toAccountNumber.toString()} for ${formatMoney(instruction.amount)} - ${result.message}`,
      );
    });
  }

  return lines.join('\n');
}
